lib/exercises.rb

```diff
@@ -2,19 +2,61 @@
 
 # This method will return an array of arrays.
 # Each subarray will have strings which are anagrams of each other
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n)
+# Space Complexity: O(n)
 
 def grouped_anagrams(strings)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
-end
-
-# This method will return the k most common elements
-# in the case of a tie it will select the first occuring element.
-# Time Complexity: ?
-# Space Complexity: ?
-def top_k_frequent_elements(list, k)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  if strings.length == 0 #empty hash if no char in string
+    return []
+  end
+  
+  anagrams_hash = {} #where its a-going to get put
+  
+  anagram_array = []
+  
+  strings.each do |word| #take each word...
+    letters = word.split("").sort.join("") #...split into letters, and rejoin them
+    if anagrams_hash[letters].nil? #if string isnt in hash
+      anagrams_hash[letters] = word
+    else 
+      anagrams_hash[letters] = [word]
+    end
+    
+    anagram_array.push(word) #shovels word in
+    
+    return anagrams_hash.values #boom boom
+    
+  end
+  
+  # This method will return the k most common elements
+  # in the case of a tie it will select the first occuring element.
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def top_k_frequent_elements(list, k)
+    return [] if list.empty? #empty array if no char in list
+  end
+  
+  common_elements_hash = {} #where its going...
+  
+  list.each do |element|
+    if common_elements_hash[element].nil? #if empty, start at one
+      common_elements_hash[element] = 1 
+    else
+      common_elements_hash[element] += 1 #otherwise increment
+    end
+  end
+  
+  elements_sort = common_elements_hash.max_by(k) {|key, value| value} #gives us k hash pairs in array form
```

Comment:
Good use of max_by

Just note that it's searching the hash of (n) elements k times.  So this is O(nk)